src/bos/server/redis_db_utils.py

```diff
@@ -106,6 +108,34 @@ def get_all(self):
             data.append(single_data)
         return data
 
+    def get_all_filtered(self,
+                         filter_func: Callable[[dict], dict | None],
+                         start_after_key: Optional[str] = None,
+                         page_size: int = 0) -> list[dict]:
+        """
+        Get an array of data for all keys after passing them through the specified filter
+        (discarding any for which the filter returns None)
+        If start_after_id is specified, all ids lexically <= that id will be skipped.
+        If page_size is specified, the list will be returned if it contains that many
+        elements, even if there may be more remaining.
+        """
+        data = []
+        for value in self.iter_values(start_after_key):
+            filtered_value = filter_func(value)
+            if filtered_value is not None:
+                data.append(filtered_value)
+                if page_size and len(data) == page_size:
+                    break
+        return data
+
+    def iter_values(self, start_after_key: Optional[str] = None):
+        all_keys = sorted({k.decode() for k in self.client.scan_iter()})
+        if start_after_key is not None:
```

Comment:
if start_after_key

Reply:
check for start_after_key 
if start_after_key >= len(all_keys):
    return
# No need to iterate through the list
if start_after_key:
    all_keys = all_keys[start_after_key+1:]

Reply:
`start_after_key` is a string. I don't see how your example code would work. Am I missing something?

Reply:
Just realized that. In that case 

if start_after_key:
all_keys = all_keys[all_keys.index(start_after_key)+1:]

Reply:
That only works if `start_after_key` is in the list, which it is not guaranteed to be. It's possibly the item got deleted, but we still want to return only items that are lexically after it in the list.